### PR TITLE
Bc 5135 activate import

### PR DESCRIPTION
**Pull request by VikDavydiuk**
# Description

QA note:
Change tested on env:
https://nbc-bc-5135-activate-import.cd.dbildungscloud.dev/
No issues found. Teacher and student work the same. Save and Open tldraw file work fine. Options: Save, Save as..., Open, works fine. Option Export as... works fine also. 
More detail: https://ticketsystem.dbildungscloud.de/browse/BC-5135

## Links to Tickets or other pull requests

- https://ticketsystem.dbildungscloud.de/browse/BC-5135

## Approval for review

- [x] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### CODE REVIEW

**Review by blazejpass — APPROVED** [review]

File: src/hooks/useMultiplayerState.ts [hunk 1]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+
+				updateYMapsInTransaction(
+					document.pages.page.shapes,
+					document.pages.page.bindings,
+					document.assets,
+				);
+
+				app.zoomToFit();
+			} catch (e) {
+				console.error(e);
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:157]: In generell:
Using console must avoid. Please introduce a logger framework/package for it. This is the sync write operations that shorty blog the node process. By high traffic it can be reduce the response time a lot.

In your case:
But your code is only executed in brower correct?

Comment (VikDavydiuk) [reply src/hooks/useMultiplayerState.ts:157]: Okay, i can add a logger package, but which one?
Loglevel? Winston?

File: src/hooks/useMultiplayerState.ts [hunk 2]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+			const reader = new FileReader();
+			reader.onloadend = () => {
+				if (reader.readyState === FileReader.DONE) {
+					resolve(reader.result as string);
+				}
+			};
+			reader.readAsText(blob, 'utf8');
+		});
+
+		const file: TDFile = JSON.parse(json);
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:91]: If you create a type/interface for 
```
{
	fileHandle: FileSystemFileHandle | null;
	document: TDDocument;
}
```

Then you can also create a builder MyBuilderX.build(json, blob) that return you the expected type and include all parsing, ?? checks and so on.

File: src/hooks/useMultiplayerState.ts [hunk 3]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+			multiple: false,
+		});
+
+		if (!blob) return null;
+
+		const json: string = await new Promise((resolve) => {
+			const reader = new FileReader();
+			reader.onloadend = () => {
+				if (reader.readyState === FileReader.DONE) {
+					resolve(reader.result as string);
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:85]: You cast without check if the type is really a string to runtime. It is possible to write ts typeguard with methodName(x: type) is Y that must return with true / false. That make the type check. In this case if it is a string. 
Than this function can be used in a cast function castToString() //1. check type // 2. cast // return with string.

File: src/hooks/useMultiplayerState.ts [hunk 4]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+	}> {
+		const blob = await fileOpen({
+			description: 'Tldraw File',
+			extensions: [`.tldr`],
+			multiple: false,
+		});
+
+		if (!blob) return null;
+
+		const json: string = await new Promise((resolve) => {
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:81]: When this is a class then the line 81 to 89 should be placed in a private method with good name what they do. 
Maybe you can outsource it in a additional function placed in scope but unsure.
This file look currently like a script. For me it is hard to follow what is going on. I think the complet file can be more cleaner and following more the best practice from react.

File: src/hooks/useMultiplayerState.ts [hunk 5]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+		};
+	}
+
+	const updateYMapsInTransaction = (
+		shapes: Record<string, TDShape | undefined>,
+		bindings: Record<string, TDBinding | undefined>,
+		assets: Record<string, TDAsset | undefined>,
+	) => {
+		doc.transact(() => {
+			Object.entries(shapes).forEach(([id, shape]) => {
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:107]: Each of this sub steps Object.entries... should be behind a good naming what is decribe what it should be do.

File: src/hooks/useMultiplayerState.ts [hunk 6]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+				} else {
+					yAssets.set(asset.id, asset);
+				}
+			});
+		});
+	};
+
+	const onMount = useCallback((app: TldrawApp) => {
+		app.loadRoom(roomId);
+		app.pause();
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:133]: This return null should it? It exists no comment that implemantion should add later. Should it work without await?
The same for loadRoom

File: src/hooks/useMultiplayerState.ts [hunk 7]
```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
...
+		setAppInstance(app);
+
+		app.openProject = async () => {
+			try {
+				const result = await openFromFileSystem();
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:138]: Without any params?

**Review by CeEv — CHANGES_REQUESTED** [review]
Maybe it make sense if we speak together. Currently for me is unclear what the PR changed.

File: src/hooks/useMultiplayerState.ts [hunk 8]
```diff
@@ -18,35 +9,44 @@ import {
...
 	yShapes,
 } from '../store/store';
+import { fileOpen } from 'browser-fs-access';
 import { TldrawPresence } from '../types';
+import { FileBuilder, FileBuilderResult } from '../utilities/fileBuilder';
+import { errorLogger } from '../utilities/logger';
+import {
+	STORAGE_SETTINGS_KEY,
+	getUserSettings,
+} from '../utilities/localStorage';
```
Comment (CeEv) [inline src/hooks/useMultiplayerState.ts:19]: Working with a barrel file / index.ts that re-export the parts in utilities make you imports more readable.
import { FileBuilder, FileBuilderResult, errorLogger, STORAGE_SETTINGS_KEY, getUserSettings } from '../utilities';

**Review by CeEv — APPROVED** [review]
